src/Accelerators/NNPA/Transform/ZLow/ZLowRewrite.cpp

```diff
@@ -100,6 +112,330 @@ class StickViewUnstickRemovalPattern : public OpRewritePattern<ZLowStickOp> {
+        NHWCIndices.emplace_back(NCHWIndices[2]); // H
+        NHWCIndices.emplace_back(NCHWIndices[3]); // W
+        NHWCIndices.emplace_back(NCHWIndices[1]); // C
+        operandMap.map(NCHWIndices, NHWCIndices);
+      }
+      Operation *clonedOp = rewriter.clone(*loadOp.getOperation(), operandMap);
+      clonedOp->getResult(0).setType(stickifiedElementType);
+      // This DummyOp is used to make the intermediate generated code valid. It
+      // wil be removed automatically via canonicalization.
+      Value dummyConverter = rewriter.create<ZLowDummyOp>(
+          loc, cpuElementType, clonedOp->getResult(0));
+      rewriter.replaceOp(loadOp, {dummyConverter});
+    }
+
+    // - Rewrite AffineStoreOp to use stickified Memref directly.
+    for (AffineStoreOp storeOp : storeOps) {
+      Value storeMemref = storeOp.getMemref();
+      Value storeValue = storeOp.getValue();
+      ZLowStickOp myStickOp = StoreOpStickOpMap[storeOp];
+      Value stickMemref = myStickOp.getOut();
+
+      // Move stickMemref's AllocOp up before affine.for so that it
+      // dominates its uses. A good place is just after storeMemref's AllocOp.
+      //
+      // Get AllocOps that allocated storeMemref and stickMemref.
+      Operation *storeAllocOp = storeMemref.getDefiningOp();
+      Operation *stickAllocOp = stickMemref.getDefiningOp();
+      // stickAllocOp should be after storeAllocOp, since dimensions come from
+      // storeAllocOp according to the definition of zlow.stick.
+      Operation *justMovedOp = nullptr;
+      // Move AllocOp's operands first.
+      for (unsigned i = 0; i < stickAllocOp->getNumOperands(); ++i) {
+        Value oprd = stickAllocOp->getOperand(i);
+        if (isa<BlockArgument>(oprd))
+          continue;
+        Operation *opToMove = oprd.getDefiningOp();
+        // Do not move, it is potentially used by storeAllocOp and it is a good
+        // place already.
+        if (opToMove->isBeforeInBlock(storeAllocOp))
+          continue;
+        if (justMovedOp)
+          opToMove->moveAfter(justMovedOp);
+        else
+          opToMove->moveAfter(storeAllocOp);
+        justMovedOp = opToMove;
+      }
+      // Move AllocOp.
+      if (justMovedOp)
+        stickAllocOp->moveAfter(justMovedOp);
+      else
+        stickAllocOp->moveAfter(storeAllocOp);
+
+      // Replace storeOp.
+      OpBuilder::InsertionGuard guard(rewriter);
+      rewriter.setInsertionPointAfter(storeOp);
+      // This DummyOp is used to make the intermediate generated code valid. It
+      // will be removed automatically via canonicalization.
+      Value dummyConverter =
+          rewriter.create<ZLowDummyOp>(loc, stickifiedElementType, storeValue);
+      // Clone storeOp with new Memref, Value, and Indices.
+      IRMapping operandMap;
+      operandMap.map(storeOp.getMemref(), stickMemref);
+      operandMap.map(storeOp.getValue(), dummyConverter);
+      // Permute indices in case of NCHW layout.
+      if (isNCHWLayout) {
```

Comment:
Don't you have to filter out more layouts initially (only saw the 1D and 2DS being filtered out). Or support more of them here?

Maybe it would be more resilient at the top if you stated the one you support, and all the ones that are not supported would then report failure...

```
if (!(layout supported)) report failure
```

Reply:
> Don't you have to filter out more layouts initially (only saw the 1D and 2DS being filtered out). Or support more of them here?

No, only 1D and 2DS are not supported since their affine_maps have issue reported here: https://github.com/onnx/onnx-mlir/issues/1940. The remaining layouts have no problem and we support them all. So I filtered out 1D and 2DS at the beginning at Line 227.